modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -276,192 +205,438 @@ private QueryPlan prepareFragment(String jsonFragment) {
 
     /** {@inheritDoc} */
     @Override
-    public SqlCursor<List<?>> executePlan(
-            RootQuery<RowT> qry,
-            QueryPlan plan
+    public AsyncCursor<List<Object>> executePlan(
+            QueryPlan plan, BaseQueryContext ctx
     ) {
         switch (plan.type()) {
             case DML:
                 // TODO a barrier between previous operation and this one
             case QUERY:
-                return mapAndExecutePlan(
-                        qry,
+                return executeQuery(
+                        ctx,
                         (MultiStepPlan) plan
                 );
             case EXPLAIN:
                 return executeExplain((ExplainPlan) plan);
             case DDL:
-                return executeDdl(qry, (DdlPlan) plan);
+                return executeDdl((DdlPlan) plan);
 
             default:
                 throw new AssertionError("Unexpected plan type: " + plan);
         }
     }
 
-    private SqlCursor<List<?>> executeDdl(RootQuery<RowT> qry, DdlPlan plan) {
+    /** Cancels the query with given id. */
+    public CompletionStage<?> cancel(UUID qryId) {
+        var mgr = queryManagerMap.get(qryId);
+
+        if (mgr == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        return mgr.close(true);
+    }
+
+    private AsyncCursor<List<Object>> executeDdl(DdlPlan plan) {
         try {
             ddlCmdHnd.handle(plan.command());
         } catch (IgniteInternalCheckedException e) {
-            throw new IgniteInternalException("Failed to execute DDL statement [stmt=" + qry.sql()
+            throw new IgniteInternalException("Failed to execute DDL statement [stmt=" /*+ qry.sql()*/
```

Comment:
Should we return here a future that completed exceptionally?

Reply:
Yes, we should. But currently DdlCommandHandler is nor refactored yet, and invocation of `org.apache.ignite.internal.sql.engine.exec.ExecutionServiceImpl#executePlan` is actually a step of CompletableFuture pipeline, so the user will get a future completed exceptionally anyway.